codeworld-server/src/Util.hs

```diff
@@ -57,6 +60,19 @@ clientIdPath = "web/clientId.txt"
 buildRootDir :: BuildMode -> FilePath
 buildRootDir (BuildMode m) = "data" </> m </> "user"
 
+androidRootDir :: BuildMode -> FilePath
+androidRootDir (BuildMode m) = "data" </> m </> "android"
+
+androidBuildDir :: BuildMode -> ProgramId -> FilePath
+androidBuildDir mode programId = androidRootDir mode </> sourceBase programId
+
+apkFile :: BuildMode -> ProgramId -> FilePath
+apkFile mode programId =
```

Comment:
Once you move the build into a temporary directory (see other comment), you will want to pick a simpler path for the apk file.

Reply:
Since you're letting people choose different app names for the same code, you cannot uniquely identify apk files by the hash of their code.  You need a different scheme.